db/view/regular_column_transformation.hh

```diff
@@ -0,0 +1,129 @@
+//    to be transformed is missing from the given row, or 3. "deleted"
+//    if the row contains a deletion (a tombstone) for the column.
+
+class regular_column_transformation : public column_computation {
+public:
+    struct result {
+        std::optional<bytes> _value;
+        bool _deleted = false; // if _deleted, "value" is unset
+        api::timestamp_type _ts = api::missing_timestamp; // defined if has_value() or is_deleted() but not if is_missing()
+        gc_clock::duration _ttl; // defined if has_value()
+        gc_clock::time_point _expiry; // defined if has_value()
+
+        // Same convention as in class row_marker
+        static constexpr gc_clock::duration no_ttl { 0 };
+        static constexpr gc_clock::time_point no_expiry { gc_clock::duration(0) };
+
+        bool has_value() const {
+            return _value.has_value();
+        }
+        // Should only be called if has_value() is true
+        const bytes& get_value() const {
+            return *_value;
+        }
+
+        // Should only be called if has_value() or is_deleted() (i.e., !is_missing())
+        api::timestamp_type get_ts() const {
+            return _ts;
+        }
+        gc_clock::duration get_ttl() const {
+            return _ttl;
+        }
+        gc_clock::time_point get_expiry() const {
+            return _expiry;
+        }
+
+        // Note: the existing code doesn't use the is_deleted() or is_missing()
+        // functions, or makes a distinction between a missing and deleted
+        // cells, so this part of this class is kind of superflous. The code
+        // using this assumes that if a cell existed prior to an update and
+        // gone after the update, it is a is_deleted() and has a deletion
+        // timestamp. If a cell didn't exist prior to the update nor after the
+        // update, we don't even care if it has a timestamp (is_missing()
+        // or is_deleted() and don't try to retrieve its timestamp)
+        bool is_deleted() const {
+            return _deleted;
+        }
+        bool is_missing() const {
+            return !has_value() && !is_deleted();
+        }
+
+        // A missing computation result
+        result() { }
+
+        // Construct a computation result by copying a given atomic_cell -
+        // including its value, timestamp, and ttl - or deletion timestamp.
+        // The second parameter is an optional transformation function f -
+        // taking a bytes and returning an optional<bytes> - that transforms
+        // the value of the cell but keeps its other liveness information.
+        // If f returns a nullopt, it means the view row should no longer
+        // exists - i.e., it's a deletion.
+        template<typename Func=std::identity>
+        result(atomic_cell_view cell, Func f = {}) {
+            static_assert(std::is_invocable_v<Func, bytes> && std::is_convertible_v<decltype(f(bytes{})), std::optional<bytes>>);
```

Comment:
I'm embarassed to admit I don't know c++20 constaints/concepts that much, but I'd guess this static_assert can be rewritten using those, perhaps, we already use concepts in our source code.

Reply:
Sadly, I tried (I thought exactly like you) - and failed. The concepts syntax is so awkward that I just failed to do it, so I took the easy way out. To be honest, it seems good enough for me :-)
Maybe I can try again.

Reply:
Don't bother, I know your pain, it's good enough for me as well ;)